DAE initialization note (#604)
Ferrite-FEM/Ferrite.jl: docs/src/manual/initial_and_boundary_conditions.md

apply_analytical!(u, dh, :p, x -> ρ * g * x[2])
```

> We refer to the paper ["Consistent Initial Condition Calculation for Differential-Algebraic Systems"  by Brown et al.](dx.doi.org/10.1137/S1064827595289996) 
> for some details on the initialization of DAEs.
> 
> !!! note "Consistency"
>     `apply_analytical!` does not enforce consistency of the applied solution with the system of 
>     equations. Some problems, like for example differential-algebraic systems of equations (DAEs)
>     need extra care during initialization, as noted above.
> 
> See also [Time Dependent Problems](@ref) for one example.

Reviewer: ```suggestion
See also [Time Dependent Problems](@ref) for one example.

!!! note "Consistency"
    `apply_analytical!` does not enforce consistency of the applied solution with the system of 
    equations. Some problems, like for example differential-algebraic systems of equations (DAEs)
    need extra care during initialization. We refer to the paper ["Consistent Initial Condition Calculation for Differential-Algebraic Systems"  by Brown et al.](dx.doi.org/10.1137/S1064827595289996) for more details on this matter.

```

apply_analytical!(u, dh, :p, x -> ρ * g * x[2])
```

> !!! note "Initialization of Differential-Algebraic Systems"
>     Differential-algebraic systems of equations (DAEs) need extra care during initialization, which is 
>     currently not automatized by the function `apply_analytical!`.
>     Skipping the formal definition, you can identify DAEs for example by 
>     * Multiple time derivatives in the equation
>     * The equations cannot be formulated such that the variable with the time-derivative 
>       is the only term on the right hand side
>     * Missing time derivatives
>     To give concrete example of a "DAE" (formally a PDAE, which becomes a DAE by spatial discretization), consider following
>     abstract problem on some domain with ``u(t,\boldsymbol{x})`` and ``s(t,\boldsymbol{x})`` being the unknown variables:
>     ```math
>         \partial_t u(t,\boldsymbol{x})) = f(u(t,\boldsymbol{x}), \nabla u(t,\boldsymbol{x}), s(t,\boldsymbol{x})) \\
>         0 = g(u(t,\boldsymbol{x}), \nabla u(t,\boldsymbol{x}), s(t,\boldsymbol{x}))
>     ```
>     Here ``s`` might be some internal state variable, e.g. describing hardening, failure or some electrical
>     state and ``u`` could be a displacement field or some potential field. This form is quite common in mechanics,
>     for example in quasi-static analysis of plastic materials.
>     For DAEs need to assert that the hidden constraints of the problem are not violated by the choice
>     of initial conditions.
>     What the exact hidden constraints are have to be worked out on the specific DAE.
>     For the example above it is sometimes possible to just solve the second equation for ``s``, i.e.
>     ```math
>     0 = g(u(t,\boldsymbol{x}), \nabla u(t,\boldsymbol{x}), s(t,\boldsymbol{x}))
>     ```
>     given some fixed choice of ``u``.
>     We refer to the paper ["Consistent Initial Condition Calculation for Differential-Algebraic Systems" 
>     by Brown et al.](dx.doi.org/10.1137/S1064827595289996) for more details on this matter and some possible solutions.
> 

Reviewer: Thanks for providing such a detailed description Dennis, it is very interesting!
But this does indeed seem very difficult to explain in simple terms for the general case to not confuse (new) users. 
Thinking more about this, could something like the following, even if over-simplistic, work?
```suggestion
!!! note "Initialization of Differential-Algebraic Systems"
    Note that `apply_analytical!` does not ensure consistency between the applied field 
    and the specific problem being solved. As a simple example, consider a mechanical problem: 
    `apply_analytical!` doesn't ensure that equilibrium is fulfilled in the initial state. 
    In general for advanced, so-called "Differential-Algebraic Systems", 
    consistent initialization is a research field on its own, see e.g. 
    ["Consistent Initial Condition Calculation for Differential-Algebraic Systems" by Brown et al.](dx.doi.org/10.1137/S1064827595289996)
    
```
or do you think this misses the point since in many cases a non-equilibrium initial condition is perfectly fine (and we even demo that in the transient heat flow:))

I am, however, sad not to have all such knowledge/insights available. From that perspective, it might have been nice to have a separate webpage containing "Ferrite: Theory and examples" (or something) to supplement the documentation, but that's a different discussion

Author: Thanks for looking into this and the suggestion. I am happy and open for ideas on how to improve/simplify this part. However, the point here is not about *equilibrium*, but about *consistency* of the solution. This is especially problematic in differential-algebraic problems. 

Think about it this way: The algebraic equations act as constraints on the time evolution of your problem, where the constraints also have some "implicit" or "hidden evolution". However, the constraint might not be directly visible in the equation and hence messing it up can be very easy (for example if a zero initialization of some internal variable could be inconsistent with your evolving field variable).

Don't you think the (theory) manual we have is sufficient in combination with the examples?

Reviewer: Yes, I agree that equilibrium is not the point. I meant it more as a warning that one can not supply whatever initial condition but have to ensure that it is consistent with the problem to be solved. 

(The additional site was more a long term idea that would be useful for e.g. teaching and reference since you, we and maybe more use Ferrite in teaching, for which I think the current state is not meant to be, nor detailed enough, as course material)

> See also [Time Dependent Problems](@ref) for one example.

Reviewer: Should we put this above the note?

Author: Quick question here, why would you move it above the note?

Reviewer: It is easy to miss that reference when reading the docs, but if switched, the note is still hard to miss. And it follows (in my mind) more logically as DAEs are not discussed in that example.

```

See also [Time Dependent Problems](@ref) for one example.

!!! note "Consistency"
    `apply_analytical!` does not enforce consistency of the applied solution with the system of 
    equations. Some problems, like for example differential-algebraic systems of equations (DAEs)

>     need extra care during initialization. We refer to the paper ["Consistent Initial Condition Calculation for Differential-Algebraic Systems"  by Brown et al.](dx.doi.org/10.1137/S1064827595289996) for more details on this matter.

Reviewer 2: Probably you should use the full https URL instead, this gives:
```
┌ Warning: invalid local link: unresolved path in manual/boundary_conditions.md
│   link_url = "dx.doi.org/10.1137/S1064827595289996"
└ @ Documenter.HTMLWriter ~/.julia/packages/Documenter/ZRco4/src/html/HTMLWriter.jl:2195
```
